### PR TITLE
Let `versionCatalogs` extension be available in precompiled Kotlin DSL scripts

### PR DESCRIPTION
**Pull request by eskatos**
* Fixes https://github.com/gradle/gradle/issues/22468

This PR also splits a large test class into several ones, hence more files changed than only necessary for the fix.
Take a look at the commits to make the review easier.

### CODE REVIEW

**Review by lkasso — APPROVED** [review]
Docs approved.

File: subprojects/docs/src/docs/release/notes.md [hunk 1]
```diff
@@ -83,6 +83,28 @@ Like the Gradle Build Tool, its Wrapper is licensed under the Apache Software Li
...
 
+<a name="kotlin-dsl"></a>
+### Kotlin DSL improvements
+
+Gradle's [Kotlin DSL](userguide/kotlin_dsl.html) provides an enhanced editing experience in supported IDEs compared to the traditional Groovy DSL — auto-completion, smart content assist, quick access to documentation, navigation to source, and context-aware refactoring.
+
+#### Version catalog API in precompiled scripts
+
+The `versionCatalogs` extension accessor is now available in Kotlin DSL precompiled scripts.
+It provides with a [type unsafe API](userguide/platforms.html#sub:type-unsafe-access-to-catalog) to access version catalogs that may be available on the projects where the precompiled script will be applied.
```
Comment (jbartok) [inline subprojects/docs/src/docs/release/notes.md:94]:
```suggestion
It provides a [type unsafe API](userguide/platforms.html#sub:type-unsafe-access-to-catalog) for accessing version catalogs available on the projects where the precompiled script will be applied.
```

File: platforms/core-configuration/kotlin-dsl-integ-tests/src/integTest/kotlin/org/gradle/kotlin/dsl/integration/PrecompiledScriptPluginVersionCatalogIntegrationTest.kt [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        build(":help").apply {
+            assertOutputContains("org.codehaus.groovy:groovy:3.0.5")
+        }
+    }
+
+    @Test
+    fun `version catalogs from outer builds are not available as accessors`() {
+        withKotlinBuildSrc()
+        withSimpleVersionCatalog()
+        val scriptBody = "println(libs)"
```
Comment (jbartok) [inline platforms/core-configuration/kotlin-dsl-integ-tests/src/integTest/kotlin/org/gradle/kotlin/dsl/integration/PrecompiledScriptPluginVersionCatalogIntegrationTest.kt:61]: I think extracting this makes the test harder to read and is worse than living with the duplication.

**Review by jbartok — APPROVED** [review]
LGTM, minor comments.

**Comment by eskatos** [comment]
@bot-gradle test and merge

**Pull request merged**